lib/puppetserver/ca/generate_action.rb

```diff
@@ -166,11 +177,46 @@ def sign_intermediate(ca_key, ca_cert, csr, valid_until, signing_digest)
           extension = ef.create_extension(*ext)
           cert.add_extension(extension)
         end
+
+        if !subject_alt_names.empty?
+          alt_names_ext = ef.create_extension("subjectAltName", subject_alt_names, false)
+          cert.add_extension(alt_names_ext)
+        end
+
         cert.sign(ca_key, signing_digest)
 
         cert
       end
 
+      # Given any alt names specified on the command line and any loaded from
+      # settings, choose the names with the highest precedence and process them.
+      # Precedence:
+      # 1) Specified on the CLI
+      # 2) Specified in settings
+      # 3) Defaults
+      def choose_alt_names(cli_alt_names, settings_alt_names)
+        if !cli_alt_names.empty?
+          sans = cli_alt_names
+        elsif !settings_alt_names.empty?
+          sans = settings_alt_names
+        else
+          sans = "puppet, #{Facter.value(:fqdn)}, puppet.#{Facter.value(:domain)}"
+        end
+        munge_alt_names(sans)
+      end
+
+      def munge_alt_names(names)
+        raw_names = names.split(/\s*,\s*/).map(&:strip)
```

Comment:
Do we have any problems with altnames having spaces?
ex:
```
"hi, yo good"
=> "hi, yo good"
irb(main):007:0> names.split(/\s*,\s*/).map(&:strip)
=> ["hi", "yo good"]
irb(main):008:0> raw_names = ["hi", "yo good"]
=> ["hi", "yo good"]
irb(main):009:0> munged_names = raw_names.map do |name|
irb(main):010:1*           # Prepend the DNS tag if no tag was specified
irb(main):011:1*           if !name.start_with?("IP:") && !name.start_with?("DNS:")
irb(main):012:2>             "DNS:#{name}"
irb(main):013:2>           else
irb(main):014:2*             name
irb(main):015:2>           end
irb(main):016:1>         end.sort.uniq.join(", ")
=> "DNS:hi, DNS:yo good"
```

Reply:
Also, if this were processed as an array (like certnames) we wouldn't have to do a split and strip, or?